### ISSUE TITLE
IWYU does not like forward-only headers

### ISSUE THREAD

**Issue by smcpeak** [include-what-you-use/include-what-you-use#1570]
In my coding style, every module with header `foo.h` also contains a forward-only header file, `foo-fwd.h`, which forward-declares all of the classes that `foo.h` defines.

But IWYU does not like this, insisting that all forward declarations be repeated in every file that makes use of them.

Minimal example:

```c++
// foo-fwd.h
// Forward-declare `class Foo`.

#ifndef FWD_FOO_H
#define FWD_FOO_H

class Foo;

#endif // FWD_FOO_H
```

```c++
// use.cc
// Use `foo-fwd.h`.

#include "foo-fwd.h"         // Foo (forward)

// Forward suffices.
Foo *p;

// EOF
```

Running IWYU:

```plaintext
$ include-what-you-use use.cc

use.cc should add these lines:
class Foo;

use.cc should remove these lines:
- #include "foo-fwd.h"  // lines 4-4

The full include-list for use.cc:
class Foo;
---

$ include-what-you-use --version
include-what-you-use 0.22 (git:1cf391d) based on clang version 18.1.4 (https://github.com/llvm/llvm-project.git e6c3289804a67ea0bb6a86fadbe454dd93b8d855)
```

The suggestion to replace the `#include` with the forward declaration it contains is counterproductive, since the whole point of `foo-fwd.h` is to have them all in one place so I [do not repeat myself](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

Second reproducer showing a problem with template default arguments:

```c++
// bar-fwd.h
// Forward declaration of class template `Bar`.

#ifndef BAR_FWD_H
#define BAR_FWD_H

template <typename T = int>
class Bar;

#endif // BAR_FWD_H
```

```c++
// use-bar.h
// Use `bar-fwd.h`.

#ifndef USE_BAR_H
#define USE_BAR_H

#include "bar-fwd.h"         // Bar (forward)

// Forward suffices.
extern Bar<> *p;

#endif // USE_BAR_H
```

Running IWYU:

```plaintext
$ include-what-you-use -xc++ use-bar.h

use-bar.h should add these lines:
template <typename T = int> class Bar;

use-bar.h should remove these lines:
- #include "bar-fwd.h"  // lines 7-7

The full include-list for use-bar.h:
template <typename T = int> class Bar;
---
```

Here, the recommendation will lead to compilation errors because copying the forward into each client will break the rule that template default arguments cannot be specified more than once in a translation unit.

This is the same core issue as in Issue #530, but that issue has an unnecessarily complicated reproducer.  Additionally, that issue has discussion of using mapping files as a solution, which I see as a wrong approach, at least for a simple case like what I've shown above, since it should not require any additional configuration for IWYU to do the right thing.  It also doesn't seem to be getting much attention despite being, for me, a deal-breaker for using IWYU.

So, I'm filing this separately with the intention of keeping this issue narrowly focused on allowing a header file that contains forwards without the need to use a mapping file, and perhaps getting it some attention, though I understand if you'd prefer to mark it as a dup anyway.

**Comment by kimgr** [comment]
Does the result change if you put `// IWYU pragma: export` after each forward decl in your fwd header?

**Comment by smcpeak** [comment]
Yes, adding `// IWYU pragma: export` to the forward declarations does work!  I had overlooked that usage of that pragma; my mistake.

For completeness, in the template example, the pragma goes on the second line; from experimentation, it's the line containing the identifier (`Bar` above).  It also works to surround the block of forwards with `// IWYU pragma: begin/end_exports`.

IMO, it's not ideal to have to add these annotations, but manually exporting them (esp. as a block) makes it feasible for me to continue using IWYU experimentally, so thanks for the tip!

Therefore, this report could be treated as an enhancement request to handle this style more automatically.  One possibility would be a configuration switch to implicitly export all forwards that are in a file whose name ends in `fwd`.

**Comment by bolshakov-a** [comment]
Default template argument part of this issue has been fixed on master branch: https://github.com/include-what-you-use/include-what-you-use/commit/0d27d9971936ef6aef87dfee7a17d37f90d4b52d
